### ISSUE TITLE
Bug/Help Needed: Private Endpoint Exposes only one NIC IP address

### ISSUE THREAD

**Issue by mehighlow** [Azure/azure-service-operator#4464]
## Describe the bug

Private Endpoint Integration for CosmosDB exposes [only one NIC IP address](https://github.com/Azure/azure-service-operator/blob/b98c83fad21dfb4d6ffcb836a6723ad73ba144ab/v2/api/network/v1api20220701/structure.txt#L1027). However, a Private Endpoint for CosmosDB might create [more than one NIC + IP address](https://learn.microsoft.com/en-us/azure/cosmos-db/how-to-configure-private-endpoints?tabs=arm-bicep#fetch-the-private-ip-addresses). 

Automation can be leveraged to create just one DNS record for `PrimaryNicPrivateIpAddress`: 
```
│   ├── OperatorSpec: *Object (3 properties)
│   │   ├── ConfigMapExpressions: *core.DestinationExpression[]
│   │   ├── ConfigMaps: *Object (1 property)
│   │   │   └── PrimaryNicPrivateIpAddress: *genruntime.ConfigMapDestination
```

although both IP addresses remain valid for establishing connections:

![Image](https://github.com/user-attachments/assets/eaeaada0-a366-4130-9834-aec73fbb2db1)

The issue arises using [CosmosDB client (Azure SDK for .NET)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.cosmosclientoptions.limittoendpoint?view=azure-dotnet). By default, client has the `LimitToEndpointProperty` set to `false`. This configuration instructs the SDK to automatically discover write and read regions and use them when the configured application region is unavailable. In my case, the application opted to use the regional FQDN, even though the endpoint connection was configured as global. This behavior can be addressed by setting the `LimitToEndpointProperty` to `true` and [ConnectionMode set to Gateway](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.connectionmode?view=azure-dotnet). But this adjustment requires knowing what to look for.

Azure Service Operator Version: 2.9.0

## Expected behavior

Both regional and global IP addresses are exposed, required FQDN records exposed.

## To Reproduce

Deploy CosmosDB with Private Endpoint integration. Use the Azure SDK for .NET to establish a connection to the database with default settings. Observe that the client selects the regional endpoint, even though the supplied database FQDN is global.

## Screenshots

N/A

## Additional context

N/A

**Comment by matthchr** [comment]
Does the `Status` of the PrivateEndpoint contain the IPs that you want?

If so, can you use the [new expressions](https://azure.github.io/azure-service-operator/guide/expressions/) feature to output all of the URLs/IPs you want to a configmap?

If `Status` does have the FQDNs you want, if you share its shape here I can help you to craft a CEL expression that will output the one(s) you want.

**Comment by mehighlow** [comment]
@matthchr , @theunrepentantgeek 

status indeed contains all IP addresses and FQDNs.

```status:
  conditions:
  - lastTransitionTime: "2024-12-18T06:03:57Z"
    observedGeneration: 1
    reason: Succeeded
    status: "True"
    type: Ready
  customDnsConfigs:
  - fqdn: dbname.documents.azure.com
    ipAddresses:
    - 10.10.16.4
  - fqdn: dbname-westus2.documents.azure.com
    ipAddresses:
    - 10.10.16.5
```

I’ll create a PR once this is done. Thanks!

**Comment by matthchr** [comment]
I don't think you need to send us a PR, you should be able to do this purely in CEL expressions in your resource now.

Here's [an example](https://playcel.undistro.io/?content=H4sIAAAAAAAAA61TTW%2FbMAz9K5x7aIPGajesO%2BQWdME%2BkG1Fh6EY5h0Yi3YE6MMT5aTpsP8%2ByjWQdECAHSr7QIrie4%2Bk9LuoyRaz4uIC7sjWwRGkAGlNcL1Ywo3FXRtD7%2FWLysuRp3tgGNCD8Yki1slsSDBWc2ZyK7uDLmwpkgbyGxODd%2BRTxqb7zoZIkqmzTdEMka1J60fe4FzwsLjvIjEbMZfo2x5bgjPhn6isZFBTwl00iWAX%2BjhIo33OWqjHQ9%2BYBmCMhCChbLPRBE2IIr7rE2hMOBUbvs8%2FLUG2P3798jnHHaYR5SYjw%2Blt70%2BHKjZoezxCji0az2lg2hOMQIux%2FpxXB2tJOic5oREEdJ0lHoVxZyLmWDV8kvzOhhVaGOwTJtsoTph6VnXPKbi3nq%2BDb0zLqjFWhnJWD%2B4UpIx6TTz6qvml%2FRSqopK1PVc61H2eAVfFZPLj8qcy3VzrXAyxuCP5LbWiBW3ln425%2FF8BxbTIHfyQeymX9QTey3jz9Xva4SMTVLmCLHpWeZD1KH108vq3iINQCVnyDPTKo6O9UoUPfSQlD2Z%2FFuBA9%2Bxwv4SXlyr%2Fb9TrY9jlllhkvXoGjivplwuapFX5df%2F5C%2FSacILlAwAA)


So something like:
```yaml
spec:
  operatorSpec:
    configMapExpressions:
    - name: my-configmap
      key: global
      value: self.status.customDnsConfigs.filter(config, matches(config.fqdn, "\w+.documents"))[0].ipAddresses[0]
    - name: my-configmap
      key: regional
      value: self.status.customDnsConfigs.filter(config, matches(config.fqdn, "\w+-\w+.documents"))[0].ipAddresses[0]
```

Note that you can avoid the `filter` + `matches` and just do `self.status.customDnsConfigs[0]` and `self.status.customDnsConfigs[1]` if you always know that there will be 2 and they will be in global followed by regional order.

If there may only be 1 (or 0), you'll need to guard against the `filter` returning nothing, which can be done like this (not beautiful I know, but it does the job):
`size(self.status.customDnsConfigs.filter(config, matches(config.fqdn, "\\w+-\\w+.documents"))) == 0 ? "" : self.status.customDnsConfigs.filter(config, matches(config.fqdn, "\\w+-\\w+.documents"))[0].ipAddresses[0]`

**Comment by matthchr** [comment]
Going to close this as I think the above should've answered your question. Feel free to reopen if it didn't.
